src/theory/bv/theory_bv_utils.cpp

```diff
@@ -26,21 +25,506 @@ namespace theory {
 namespace bv {
 namespace utils {
 
-Node mkSum(std::vector<Node>& children, unsigned width)
+/* ------------------------------------------------------------------------- */
+
+uint32_t pow2(uint32_t n)
+{
+  Assert(n < 32);
+  uint32_t one = 1;
+  return one << n;
+}
+
+/* ------------------------------------------------------------------------- */
+
+BitVector mkBitVectorOnes(unsigned size)
+{
+  Assert(size > 0);
+  return BitVector(1, Integer(1)).signExtend(size - 1);
+}
+
+BitVector mkBitVectorMinSigned(unsigned size)
+{
+  Assert(size > 0);
+  return BitVector(size).setBit(size - 1);
+}
+
+BitVector mkBitVectorMaxSigned(unsigned size)
+{
+  Assert(size > 0);
+  return ~mkBitVectorMinSigned(size);
+}
+
+/* ------------------------------------------------------------------------- */
+
+unsigned getSize(TNode node)
+{
+  return node.getType().getBitVectorSize();
+}
+
+const bool getBit(TNode node, unsigned i)
+{
+  Assert(i < utils::getSize(node) && node.getKind() == kind::CONST_BITVECTOR);
+  Integer bit = node.getConst<BitVector>().extract(i, i).getValue();
+  return (bit == 1u);
+}
+
+/* ------------------------------------------------------------------------- */
+
+unsigned getExtractHigh(TNode node)
+{
+  return node.getOperator().getConst<BitVectorExtract>().high;
+}
+
+unsigned getExtractLow(TNode node)
+{
+  return node.getOperator().getConst<BitVectorExtract>().low;
+}
+
+unsigned getSignExtendAmount(TNode node)
+{
+  return node.getOperator().getConst<BitVectorSignExtend>().signExtendAmount;
+}
+
+/* ------------------------------------------------------------------------- */
+
+bool isZero(TNode node)
+{
+  if (!node.isConst()) return false;
+  return node == utils::mkConst(utils::getSize(node), 0u);
+}
+
+unsigned isPow2Const(TNode node, bool& isNeg)
+{
+  if (node.getKind() != kind::CONST_BITVECTOR)
+  {
+    return false;
+  }
+
+  BitVector bv = node.getConst<BitVector>();
+  unsigned p = bv.isPow2();
+  if (p != 0)
+  {
+    isNeg = false;
+    return p;
+  }
+  BitVector nbv = -bv;
+  p = nbv.isPow2();
+  if (p != 0)
+  {
+    isNeg = true;
+    return p;
+  }
+  return false;
+}
+
+bool isBVGroundTerm(TNode node)
+{
+  if (node.getNumChildren() == 0)
+  {
+    return node.isConst();
+  }
+
+  for (size_t i = 0; i < node.getNumChildren(); ++i)
+  {
+    if (!node[i].isConst())
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool isBVPredicate(TNode node)
+{
+  if (node.getKind() == kind::EQUAL || node.getKind() == kind::BITVECTOR_ULT
```

Comment:
Nit: this could be simply written as `return (...);` instead of the if statement. (Not sure if it makes sense to address this in this PR)